src/shared/mount-setup.c

```diff
@@ -205,27 +205,26 @@ static int mount_one(const MountPoint *p, bool relabel) {
                         (void) umount2(p->where, UMOUNT_NOFOLLOW);
                         (void) rmdir(p->where);
 
-                        log_full_errno(priority, r, "Mount point %s not writable after mounting: %m", p->where);
+                        log_full_errno(priority, r, "Mount point %s not writable after mounting, undoing: %m", p->where);
                         return (p->mode & MNT_FATAL) ? r : 0;
                 }
         }
 
         return 1;
 }
 
-static int mount_points_setup(unsigned n, bool loaded_policy) {
-        unsigned i;
-        int r = 0;
+static int mount_points_setup(size_t n, bool loaded_policy) {
+        int ret = 0, r;
 
-        for (i = 0; i < n; i ++) {
-                int j;
+        assert(n <= ELEMENTSOF(mount_table));
 
-                j = mount_one(mount_table + i, loaded_policy);
-                if (j != 0 && r >= 0)
-                        r = j;
+        FOREACH_ARRAY(mp, mount_table, n) {
+                r = mount_one(mp, loaded_policy);
+                if (r != 0 && ret >= 0)
+                        ret = r;
         }
 
-        return r;
+        return ret;
 }
 
 int mount_setup_early(void) {
```

Comment:
The `assert_cc` below is redundant now?

Reply:
hmm? what do you mean? the new assert() in mount_points_setup() just checks if the specific invocation makes sense, while the new assert_cc() the patch adds just checks that the definition of N_EARLY_MOUNTS roughly makes sense. That's all.

I mean, all asserts are by definition redundant, they are just safeguards and encode assumptions. While the assert() and the assert_cc() here might check similar things, there's still quite a major semantical difference between "check a definition for validity" and "check a function invocation for validity".

Reply:
I meant you can remove https://github.com/systemd/systemd/pull/27647/files#diff-167b86cba7662f37f02f27530b191527b457bb8755c65163bd167fded58ffc49R231 :)

It's the same as the newly-added `assert_cc`